<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1.0"><title>Fit a generalized linear model for differential expression — fit_linear_model • rdevil</title><!-- jquery --><script src="https://cdnjs.cloudflare.com/ajax/libs/jquery/3.4.1/jquery.min.js" integrity="sha256-CSXorXvZcTkaix6Yvo6HppcZGetbYMGWSFlBw8HfCJo=" crossorigin="anonymous"></script><!-- Bootstrap --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/css/bootstrap.min.css" integrity="sha256-bZLfwXAP04zRMK2BjiO8iu9pf4FbLqX6zitd+tIvLhE=" crossorigin="anonymous"><script src="https://cdnjs.cloudflare.com/ajax/libs/twitter-bootstrap/3.4.1/js/bootstrap.min.js" integrity="sha256-nuL8/2cJ5NDSSwnKD8VqreErSWHtnEP9E7AySL+1ev4=" crossorigin="anonymous"></script><!-- bootstrap-toc --><link rel="stylesheet" href="../bootstrap-toc.css"><script src="../bootstrap-toc.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet"><script src="../pkgdown.js"></script><meta property="og:title" content="Fit a generalized linear model for differential expression — fit_linear_model"><meta property="og:description" content="Fit a generalized linear model in order to
model the expression of genes coming from a scRNA count matrix.
The resulting object retains the inferred parameters that can
later be tested in order to find differentially expressed genes."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body data-spy="scroll" data-target="#toc">
    

    <div class="container template-reference-topic">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar" aria-expanded="false">
        <span class="sr-only">Toggle navigation</span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <span class="navbar-brand">
        <a class="navbar-link" href="../index.html">rdevil</a>
        <span class="version label label-default" data-toggle="tooltip" data-placement="bottom" title="">0.1.0</span>
      </span>
    </div>

    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav"><li>
  <a href="../reference/index.html">Reference</a>
</li>
      </ul><ul class="nav navbar-nav navbar-right"><li>
  <a href="https://github.com/caravagnalab/rdevil/" class="external-link">
    <span class="fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->

      

      </header><div class="row">
  <div class="col-md-9 contents">
    <div class="page-header">
    <h1>Fit a generalized linear model for differential expression</h1>
    <small class="dont-index">Source: <a href="https://github.com/caravagnalab/rdevilR/fit_linear_model.R" class="external-link"><code>R/fit_linear_model.R</code></a></small>
    <div class="hidden name"><code>fit_linear_model.Rd</code></div>
    </div>

    <div class="ref-description">
    <p>Fit a generalized linear model in order to
model the expression of genes coming from a scRNA count matrix.
The resulting object retains the inferred parameters that can
later be tested in order to find differentially expressed genes.</p>
    </div>

    <div id="ref-usage">
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">fit_linear_model</span><span class="op">(</span></span>
<span>  <span class="va">input_matrix</span>,</span>
<span>  <span class="va">model_matrix</span>,</span>
<span>  <span class="va">ncounts</span>,</span>
<span>  group_matrix <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_specific_model_tensor <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  kernel_input <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  gene_names <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  cell_names <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  inference_method <span class="op">=</span> <span class="st">"SVI"</span>,</span>
<span>  method_specific_args <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div id="arguments">
    <h2>Arguments</h2>
    <dl><dt>input_matrix</dt>
<dd><p>Matrix of counts representing gene expression
data for individual cells. Each row corresponds to a gene, and each
column represents a single cell.</p></dd>


<dt>model_matrix</dt>
<dd><p>Matrix also known as design matrix, it represents
the relationship between the response variable and the predictor
variables in the model. Each row represents a cell and each columns
represent a predictor variable (e.g experimental conditions, biological
factors, treatment groups, batch effects, ...)</p></dd>


<dt>ncounts</dt>
<dd><p>Vector with length equal to the number of cells. It
contains some kind of factor relate to the library size. For example,
it might contain the total UMI of each cell (i.e. colSums(input_matrix))
or the UMI of each cell divided by the average across all of them.</p></dd>


<dt>gene_names</dt>
<dd><p>Vector containing the names of the genes</p></dd>


<dt>cell_names</dt>
<dd><p>Vector containing the names of the cells</p></dd>


<dt>inference_method</dt>
<dd><p>String. Either "SVI" or "HMC"</p></dd>


<dt>method_specific_args</dt>
<dd><p>List containing additional arguments.
The available arguments differs between the inference algorithms.</p>
<p>SVI only:</p><ul><li><p><code>optimizer_name</code> optimizer, one of "ClippedAdam", "Adam", and "SGD";</p></li>
<li><p><code>steps</code> number of iterations of the optimization algorithm;</p></li>
<li><p><code>lr</code> learning rate for the optimize;</p></li>
<li><p><code>gamma_lr</code> parameters to tune the decay of the learning rate using "ClippedAdam";</p></li>
<li><p><code>batch_size</code> number of data points or observations sampled from the input matrix in
each iteration of the optimization algorithm;</p></li>
</ul><p>HMC only:</p><ul><li><p><code>num_samples</code> number of iterations after the warmup-phase, it also indicates the
posterior samples each chain will produce;</p></li>
<li><p><code>num_chains</code> number of chains for the optimization algorithm;</p></li>
<li><p><code>warmup_steps</code> number of iterations of the warmup-phase;</p></li>
</ul><p>Shared:</p><ul><li><p><code>cuda</code> ;</p></li>
<li><p><code>jit_compile</code> ;</p></li>
<li><p><code>full_cov</code> ;</p></li>
<li><p><code>theta_bounds</code> ;</p></li>
<li><p><code>init_loc</code> ;</p></li>
</ul></dd>

</dl></div>

  </div>
  <div class="col-md-3 hidden-xs hidden-sm" id="pkgdown-sidebar">
    <nav id="toc" data-toggle="toc" class="sticky-top"><h2 data-toc-skip>Contents</h2>
    </nav></div>
</div>


      <footer><div class="copyright">
  <p></p><p>Developed by Giovanni Santacatterina.</p>
</div>

<div class="pkgdown">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

      </footer></div>

  


  

  </body></html>

